src/levanter/eval_harness.py

```diff
@@ -0,0 +1,564 @@
+#
+#             # Append metrics
+#             bpb_list.append(correct_bpb)
+#             logprob_list.append(correct_logprob)
+#             choice_logprob_list.append(correct_choice_logprob)
+#             choice_prob_norm_list.append(correct_choice_prob_norm)
+#
+#         # Aggregate metrics for the task
+#         metrics[task] = {
+#             "bpb": np.mean(bpb_list) if bpb_list else 0.0,
+#             "logprob": np.mean(logprob_list) if logprob_list else 0.0,
+#             "choice_logprob": np.mean(choice_logprob_list) if choice_logprob_list else 0.0,
+#             "choice_prob_norm": np.mean(choice_prob_norm_list)  if choice_prob_norm_list else 0.0,
+#         }
+#
+#     return metrics
+
+
+def run_eval_harness_main(config: EvalHarnessMainConfig):
+    config.trainer.initialize()
+    tokenizer = config.the_tokenizer
+
+    compute_axis_mapping = config.trainer.compute_axis_mapping
+    parameter_axis_mapping = config.trainer.parameter_axis_mapping
+
+    with config.trainer.device_mesh, hax.axis_mapping(parameter_axis_mapping):
+        key = jax.random.PRNGKey(0)
+
+        vocab_size = len(tokenizer)
+        Vocab = round_axis_for_partitioning(hax.Axis("vocab", vocab_size), compute_axis_mapping)
+        if vocab_size != Vocab.size:
+            logger.info(f"Rounding vocab size from {vocab_size} to {Vocab.size} for partitioning")
+
+        model: LmHeadModel
+
+        # initialize the model
+        if config.checkpoint_is_hf:
+            model_config = config.model
+            converter: HFCheckpointConverter = model_config.hf_checkpoint_converter()
+            converter = converter.replaced(reference_checkpoint=config.checkpoint_path, tokenizer=tokenizer)
+            model = converter.load_pretrained(
+                model_config.model_type, ref=config.checkpoint_path, dtype=config.trainer.mp.compute_dtype  # type: ignore
+            )
+        else:
+            with use_cpu_device():
+                model = eqx.filter_eval_shape(config.model.build, Vocab, key=key)
+                model = load_checkpoint(model, config.checkpoint_path, subpath="model")
+            model = hax.shard(model, parameter_axis_mapping)
+
+        model = typing.cast(LmHeadModel, inference_mode(model, True))
+
+        logger.info("Running LM eval harness....")
+        outputs = run_lm_eval_harness(
+            config.eval_harness,
+            model,
+            tokenizer,
+            config.EvalBatch,
+            axis_resources=compute_axis_mapping,
+        )
+
+        logger.info("Finished running LM eval harness")
+        # log the results as json
+        with open("lm_eval_results.json", "w") as f:
+            json.dump(outputs, f, indent=2)
+
+        # also write to stdout
+        if jax.process_index() == 0:
+            print(json.dumps(outputs, indent=2), flush=True)
+
+        # also log the results
+        levanter.tracker.current_tracker().log_artifact("lm_eval_results.json")
```

Comment:
update the filename (I'd like lm_eval_harness to be a collocation that we use consistently)